cmd/preferredimports/preferredimports.go

```diff
@@ -0,0 +1,202 @@
+
+var (
+	importAliases = flag.String("import-aliases", "hack/.import-aliases", "json file with import aliases")
+	isTerminal = terminal.IsTerminal(int(os.Stdout.Fd()))
+	logPrefix = ""
+	aliases           map[string]string
+)
+
+type analyzer struct {
+	fset      *token.FileSet // positions are relative to fset
+	ctx       build.Context
+	failed    bool
+	donePaths map[string]interface{}
+	errors    []string
+}
+
+func newAnalyzer() *analyzer {
+	ctx := build.Default
+	ctx.CgoEnabled = true
+
+	a := &analyzer{
+		fset:      token.NewFileSet(),
+		ctx:       ctx,
+		donePaths: make(map[string]interface{}),
+	}
+
+	return a
+}
+
+// collect extracts test metadata from a file.
+func (a *analyzer) collect(dir string) {
+	if _, ok := a.donePaths[dir]; ok {
+		return
+	}
+	a.donePaths[dir] = nil
+
+	// Create the AST by parsing src.
+	fs, err := parser.ParseDir(a.fset, dir, nil, parser.AllErrors)
+
+	if err != nil {
+		fmt.Println("ERROR(syntax)", logPrefix, err)
+		a.failed = true
+		return
+	}
+
+	for _, p := range fs {
+		// returns first error, but a.handleError deals with it
+		files := a.filterFiles(p.Files)
+		for _, file := range files {
+			for _, imp := range file.Imports {
+				importPath := strings.Replace(imp.Path.Value, "\"", "", -1)
+				importName := ""
+				if imp.Name != nil {
+					importName = imp.Name.Name
+				}
+				if alias, ok := aliases[importPath]; ok {
+					if alias != importName {
+						_, _ = fmt.Fprintf(os.Stderr, "%sERROR wrong alias for import \"%s\" should be %s in directory %s\n", logPrefix, importPath, alias, dir)
+						a.failed = true
+						return
+					}
+				}
+			}
+		}
+	}
+}
+
+// filterFiles restricts a list of files to only those that should be built by
+// the current platform. This includes both build suffixes (_windows.go) and build
+// tags ("// +build !linux" at the beginning).
+func (a *analyzer) filterFiles(fs map[string]*ast.File) []*ast.File {
+	files := []*ast.File{}
+	for _, f := range fs {
+		fpath := a.fset.File(f.Pos()).Name()
+		dir, name := filepath.Split(fpath)
+		matches, err := a.ctx.MatchFile(dir, name)
+		if err != nil {
+			_, _ = fmt.Fprintf(os.Stderr, "%sERROR reading %s: %s\n", logPrefix, fpath, err)
+			a.failed = true
+			continue
+		}
+		if matches {
+			files = append(files, f)
+		}
+	}
+	return files
+}
+
+type collector struct {
+	dirs []string
+}
+
+// handlePath walks the filesystem recursively, collecting directories,
+// ignoring some unneeded directories (hidden/vendored) that are handled
+// specially later.
+func (c *collector) handlePath(path string, info os.FileInfo, err error) error {
+	if err != nil {
+		return err
+	}
+	if info.IsDir() {
+		// Ignore hidden directories (.git, .cache, etc)
+		if len(path) > 1 && path[0] == '.' ||
+			// Staging code is symlinked from vendor/k8s.io, and uses import
+			// paths as if it were inside of vendor/. It fails typechecking
+			// inside of staging/, but works when typechecked as part of vendor/.
+			path == "staging" ||
```

Comment:
an ignore liked file would be better

```go
.*
staging/
vendor/
_output/
pkg/kubectl/cmd/testdata/edit
```

Reply:
Right, i *really* don't want to add more files, so having it in code forces that.